cmd/ovirt-populator/main.go

```diff
@@ -0,0 +1,262 @@
+	flag.StringVar(&metricsPath, "metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+	// Other args
+	flag.BoolVar(&showVersion, "version", false, "display the version string")
+	flag.StringVar(&namespace, "namespace", "konveyor-forklift", "Namespace to deploy controller")
+	flag.Parse()
+
+	if showVersion {
+		fmt.Println(os.Args[0], version)
+		os.Exit(0)
+	}
+
+	switch mode {
+	case "controller":
+		var (
+			gk  = schema.GroupKind{Group: groupName, Kind: kind}
+			gvr = schema.GroupVersionResource{Group: groupName, Version: apiVersion, Resource: resource}
+		)
+		populator_machinery.RunController(masterURL, kubeconfig, imageName, httpEndpoint, metricsPath,
+			namespace, prefix, gk, gvr, mountPath, devicePath, getPopulatorPodArgs)
+	case "populate":
+		populate(masterURL, kubeconfig, crName, engineUrl, secretName, diskID, fileName, namespace, crNamespace)
+	default:
+		klog.Fatalf("Invalid mode: %s", mode)
+	}
+}
+
+type engineConfig struct {
+	URL      string
+	username string
+	password string
+	ca       string
+}
+
+type TransferProgress struct {
+	Transferred uint64  `json:"transferred"`
+	Description string  `json:"description"`
+	Size        *uint64 `json:"size,omitempty"`
+	Elapsed     float64 `json:"elapsed"`
+}
+
+func getSecret(secretName, engineURL, namespace string) engineConfig {
+	config, err := rest.InClusterConfig()
+	if err != nil {
+		klog.Fatal(err.Error())
+	}
+	// creates the clientset
+	clientset, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		klog.Fatal(err.Error())
+	}
+
+	secret, err := clientset.CoreV1().Secrets(namespace).Get(context.TODO(), secretName, metav1.GetOptions{})
+	if err != nil {
+		klog.Fatal(err.Error())
+	}
+
+	return engineConfig{
+		URL:      engineURL,
+		username: string(secret.Data["user"]),
+		password: string(secret.Data["password"]),
+		ca:       string(secret.Data["cacert"]),
+	}
+}
+
+func getPopulatorPodArgs(rawBlock bool, u *unstructured.Unstructured) ([]string, error) {
+	var ovirtVolumePopulator v1beta1.OvirtVolumePopulator
+	err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.UnstructuredContent(), &ovirtVolumePopulator)
+	args := []string{"--mode=populate"}
+	if nil != err {
+		return nil, err
+	}
+
+	if rawBlock {
+		args = append(args, "--file-name="+devicePath)
+	} else {
+		args = append(args, "--file-name="+mountPath+"disk.img")
+	}
+
+	args = append(args, "--secret-name="+ovirtVolumePopulator.Spec.EngineSecretName)
+	args = append(args, "--disk-id="+ovirtVolumePopulator.Spec.DiskID)
+	args = append(args, "--engine-url="+ovirtVolumePopulator.Spec.EngineURL)
+	args = append(args, "--cr-name="+ovirtVolumePopulator.Name)
+	args = append(args, "--cr-namespace="+ovirtVolumePopulator.Namespace)
+
+	return args, nil
+}
+
+func populate(masterURL, kubeconfig, crName, engineURL, secretName, diskID, fileName, namespace, crNamespace string) {
```

Comment:
can we add a timeout setting for disk transfers?

Reply:
TBH, i'm not sure that will be good. if the transfer being executed - would you like to suddenly hit a timeout? might be 1 byte left to transfer. also, it depends on network and size.
@bennyz what do you think?

Reply:
we have a 1 minute timeout (for idle requests) in imageio already, and ovirt-engine's limit is 50 hours which is fine as the transfer of a very large disk can theoretically take over day, so I don't think we need to add a timeout on our end as well, unless we find an actual issue